### PR TITLE
Feature/core 824 data handling

### PR DESCRIPTION
**Pull request by Tschuck**

### CODE REVIEW

**Comment by datrinh** [comment]
There is a lot of duplication between DAppTwin.ts and DAppContainer. Is this avoidable?

File: dapps/digital-twin-detail/src/TwinDAppComponent.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+// intenral imports
+import DAppTwin from './DAppTwin';
+import DAppContainer from './DAppContainer';
+
+/**
+ * Used to handle correct typings for the twin detail dapp.
+ */
+@Component
+export default class TwinDAppComponent extends mixins(EvanComponent) {
```
Comment (datrinh) [inline dapps/digital-twin-detail/src/TwinDAppComponent.ts]: I would avoid another inheritance hierarchy if possible. Instead we should rework our vuex store (or wait until it works better in vue3). If the store is available in all EvanComponents. then we could just put it there?

File: dapps/digital-twin-detail/src/components/DigitalTwinDetail.ts [hunk 2]
```diff
@@ -23,8 +23,22 @@ import Component, { mixins } from 'vue-class-component';
 // evan.network imports
 import { EvanComponent } from '@evan.network/ui-vue-core';
 
+// intenral imports
```
Comment (datrinh) [inline dapps/digital-twin-detail/src/components/DigitalTwinDetail.ts]:
```suggestion
// internal imports
```

File: dapps/digital-twin-detail/src/components/DigitalTwinDetail.ts [hunk 3]
```diff
@@ -23,8 +23,22 @@ import Component, { mixins } from 'vue-class-component';
...
 @Component
-export default class DigitalTwinDetailComponent extends mixins(EvanComponent) {
+export default class DigitalTwinDetailComponent extends mixins(TwinDAppComponent) {
+  /**
+   * Show loading symbol
+   */
+  loading = true;
+
+  /**
+   * Watch for hash updates and load digitaltwin detail, if a digitaltwin was laod
```
Comment (datrinh) [inline dapps/digital-twin-detail/src/components/DigitalTwinDetail.ts]:
```suggestion
   * Watch for hash updates and load digitaltwin detail, if a digitaltwin was loaded
```

File: dapps/digital-twin-detail/src/components/DigitalTwinDetail.ts [hunk 4]
```diff
@@ -54,7 +68,43 @@ export default class DigitalTwinDetailComponent extends mixins(EvanComponent) {
...
+    this.hashChangeWatcher && window.removeEventListener('hashchange', this.hashChangeWatcher);
+    this.$store.state.twin && this.$store.state.twin.stopWatchDispatchers();
+  }
+
   close() {
     window.location.hash = `/${this.dapp.rootEns}/assets.${this.dapp.domainName}/digitaltwins`;
   }
+
+  /**
+   * Setup digital twin functionallities.
```
Comment (datrinh) [inline dapps/digital-twin-detail/src/components/DigitalTwinDetail.ts]:
```suggestion
   * Setup digital twin functionalities.
```

File: dapps/digital-twin-detail/src/components/DigitalTwinDetail.ts [hunk 5]
```diff
@@ -54,7 +68,43 @@ export default class DigitalTwinDetailComponent extends mixins(EvanComponent) {
...
+   */
+  async initialize() {
+    let beforeTwin;
+    let beforeContainer;
```
Comment (datrinh) [inline dapps/digital-twin-detail/src/components/DigitalTwinDetail.ts]: This is unused?

**Comment by datrinh** [comment]
We need to find a consensus on how to use the $store. Currently it's basically a global variable and not really vuex.

File: dapps/digital-twin-detail/src/dispatchers/container/save.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+  along with this program. If not, see http://www.gnu.org/licenses/ or
+  write to the Free Software Foundation, Inc., 51 Franklin Street,
+  Fifth Floor, Boston, MA, 02110-1301 USA, or download the license from
+  the following URL: https://evan.network/license/
+*/
+
+import * as dappBrowser from '@evan.network/ui-dapp-browser';
+import * as bcc from '@evan.network/api-blockchain-core';
+import { Dispatcher, DispatcherInstance, deepEqual, cloneDeep } from '@evan.network/ui';
+import { utils } from '@evan.network/digitaltwin.lib';
```
Comment (datrinh) [inline dapps/digital-twin-detail/src/dispatchers/container/save.ts]:
```suggestion
```

File: dapps/digital-twin-detail/src/dispatchers/container/save.ts [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    // be able to save already saved entries to indexedDB, so we will not save duplicated on errors
+    data.entriesToSave = data.entriesToSave || Object.keys(data.value);
+  })
+  .step(async (instance: DispatcherInstance, data: any) => {
+    const container = new bcc.Container(instance.runtime as bcc.ContainerOptions, data.address);
+    const { template } = await container.toPlugin();
+
+    // copy the entries to save, so the iteration will not be affected by removing entries to save
+    // from the data object => entries will be removed and the data will be persisted, after the
+    // synchronization of this entry was saved successfull, so the user won't do this twice
```
Comment (datrinh) [inline dapps/digital-twin-detail/src/dispatchers/container/save.ts]:
```suggestion
    // synchronization of this entry was saved successful, so the user won't do this twice
```

File: dapps/digital-twin-detail/src/dispatchers/container/save.ts [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+  You should have received a copy of the GNU Affero General Public License
+  along with this program. If not, see http://www.gnu.org/licenses/ or
+  write to the Free Software Foundation, Inc., 51 Franklin Street,
+  Fifth Floor, Boston, MA, 02110-1301 USA, or download the license from
+  the following URL: https://evan.network/license/
+*/
+
+import * as dappBrowser from '@evan.network/ui-dapp-browser';
+import * as bcc from '@evan.network/api-blockchain-core';
+import { Dispatcher, DispatcherInstance, deepEqual, cloneDeep } from '@evan.network/ui';
```
Comment (datrinh) [inline dapps/digital-twin-detail/src/dispatchers/container/save.ts]:
```suggestion
import { Dispatcher, DispatcherInstance, cloneDeep } from '@evan.network/ui';
```

File: dapps/digital-twin-detail/src/dispatchers/container/share.ts [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+  along with this program. If not, see http://www.gnu.org/licenses/ or
+  write to the Free Software Foundation, Inc., 51 Franklin Street,
+  Fifth Floor, Boston, MA, 02110-1301 USA, or download the license from
+  the following URL: https://evan.network/license/
+*/
+
+import * as dappBrowser from '@evan.network/ui-dapp-browser';
+import * as bcc from '@evan.network/api-blockchain-core';
+import { Dispatcher, DispatcherInstance } from '@evan.network/ui';
+import { utils } from '@evan.network/digitaltwin.lib';
```
Comment (datrinh) [inline dapps/digital-twin-detail/src/dispatchers/container/share.ts]:
```suggestion
```

File: dapps/digital-twin-detail/src/dispatchers/description.ts [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+  See the GNU Affero General Public License for more details.
+
+  You should have received a copy of the GNU Affero General Public License
+  along with this program. If not, see http://www.gnu.org/licenses/ or
+  write to the Free Software Foundation, Inc., 51 Franklin Street,
+  Fifth Floor, Boston, MA, 02110-1301 USA, or download the license from
+  the following URL: https://evan.network/license/
+*/
+
+import * as bcc from '@evan.network/api-blockchain-core';
```
Comment (datrinh) [inline dapps/digital-twin-detail/src/dispatchers/description.ts]:
```suggestion
```

File: dapps/digital-twin-detail/src/dispatchers/description.ts [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+import * as dappBrowser from '@evan.network/ui-dapp-browser';
+import { Dispatcher, DispatcherInstance } from '@evan.network/ui';
+import { EvanComponent, EvanForm, EvanFormControl } from '@evan.network/ui-vue-core';
```
Comment (datrinh) [inline dapps/digital-twin-detail/src/dispatchers/description.ts]:
```suggestion
```

File: dapps/digital-twin-detail/src/dispatchers/description.ts [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+import { EvanUIDigitalTwin } from '@evan.network/digitaltwin.lib'
```
Comment (datrinh) [inline dapps/digital-twin-detail/src/dispatchers/description.ts]:
```suggestion
```

File: dapps/digital-twin-detail/src/dispatchers/description.ts [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+import { utils } from '@evan.network/digitaltwin.lib';
```
Comment (datrinh) [inline dapps/digital-twin-detail/src/dispatchers/description.ts]:
```suggestion
```

File: dapps/digital-twin-detail/src/dispatchers/twin/favorite.add.ts [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+  along with this program. If not, see http://www.gnu.org/licenses/ or
+  write to the Free Software Foundation, Inc., 51 Franklin Street,
+  Fifth Floor, Boston, MA, 02110-1301 USA, or download the license from
+  the following URL: https://evan.network/license/
+*/
+
+import * as bcc from '@evan.network/api-blockchain-core';
+import * as dappBrowser from '@evan.network/ui-dapp-browser';
+import { Dispatcher, DispatcherInstance } from '@evan.network/ui';
+import { EvanComponent, EvanForm, EvanFormControl } from '@evan.network/ui-vue-core';
```
Comment (datrinh) [inline dapps/digital-twin-detail/src/dispatchers/twin/favorite.add.ts]:
```suggestion
```

File: dapps/digital-twin-detail/src/dispatchers/twin/favorite.add.ts [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+import { EvanUIDigitalTwin, utils } from '@evan.network/digitaltwin.lib'
```
Comment (datrinh) [inline dapps/digital-twin-detail/src/dispatchers/twin/favorite.add.ts]:
```suggestion
```

File: dapps/digital-twin-detail/src/dispatchers/twin/favorite.remove.ts [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+  along with this program. If not, see http://www.gnu.org/licenses/ or
+  write to the Free Software Foundation, Inc., 51 Franklin Street,
+  Fifth Floor, Boston, MA, 02110-1301 USA, or download the license from
+  the following URL: https://evan.network/license/
+*/
+
+import * as bcc from '@evan.network/api-blockchain-core';
+import * as dappBrowser from '@evan.network/ui-dapp-browser';
+import { Dispatcher, DispatcherInstance } from '@evan.network/ui';
+import { EvanComponent, EvanForm, EvanFormControl } from '@evan.network/ui-vue-core';
```
Comment (datrinh) [inline dapps/digital-twin-detail/src/dispatchers/twin/favorite.remove.ts]:
```suggestion
```

File: dapps/digital-twin-detail/src/dispatchers/twin/favorite.remove.ts [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+import { EvanUIDigitalTwin, utils } from '@evan.network/digitaltwin.lib'
```
Comment (datrinh) [inline dapps/digital-twin-detail/src/dispatchers/twin/favorite.remove.ts]:
```suggestion
```

File: dapps/digital-twin-detail/src/TwinDAppComponent.ts (hunk 1 above)
Comment (Tschuck) [reply dapps/digital-twin-detail/src/TwinDAppComponent.ts]: Would not prefer to move this to the global store. In this case, it's more a specific store for this DApp. I can try to remove the inheritance by adjusting the vue initialization, to use a different vuex store interface for this dapp.

**Pull request merged**